subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishHttpIntegTest.groovy

```diff
@@ -342,16 +319,70 @@ class MavenPublishHttpIntegTest extends AbstractMavenPublishIntegTest {
         module.assertPublishedAsJavaModule()
     }
 
-    private String publicationBuild(String version, String group, URI uri, PasswordCredentials credentials = null) {
+    @ToBeFixedForInstantExecution
+    def "can publish to authenticated repository using Provider to supply credentials"() {
+        given:
+        String credentialsBlock = """
+            credentials(project.credentials.usernameAndPassword('mavenRepo'))
+        """
+        buildFile << publicationBuild(version, group, mavenRemoteRepo.uri, credentialsBlock)
+        server.authenticationScheme = AuthScheme.BASIC
+        PasswordCredentials credentials = new DefaultPasswordCredentials('username', 'password')
+        expectPublishModuleWithCredentials(module, credentials)
+
+        when:
+        executer.withArguments("-PmavenRepoUsername=${credentials.username}", "-PmavenRepoPassword=${credentials.password}")
+        succeeds 'publish'
+
+        then:
+        def localPom = file("build/publications/maven/pom-default.xml").assertIsFile()
+        def localArtifact = file("build/libs/publish-2.jar").assertIsFile()
+        module.pomFile.assertIsCopyOf(localPom)
+        module.pom.verifyChecksums()
+        module.artifactFile.assertIsCopyOf(localArtifact)
+        module.artifact.verifyChecksums()
+        module.moduleMetadata.verifyChecksums()
+        module.rootMetaData.verifyChecksums()
+        module.rootMetaData.versions == ["2"]
+    }
+
+    @ToBeFixedForInstantExecution
+    def "fails at configuration time with helpful error message when username and password provider has no value"() {
+        given:
+        String credentialsBlock = """
+            credentials(project.credentials.usernameAndPassword('mavenRepo'))
+        """
+        buildFile << publicationBuild(version, group, mavenRemoteRepo.uri, credentialsBlock)
+
+        when:
+        fails 'publish'
+
+        then:
+        notExecuted(':jar', ':publishMavenPublicationToMavenRepository')
+        failure.assertHasDescription("Could not determine the dependencies of task ':publish'.")
+        failure.assertHasCause("Could not create task ':publishMavenPublicationToMavenRepository'.")
+        failure.assertHasCause("Cannot query the value of username and password provider because it has no value available")
+        failure.assertHasErrorOutput("The value of this provider is derived from")
+        failure.assertHasErrorOutput("- Gradle property 'mavenRepoUsername'")
+        failure.assertHasErrorOutput("- Gradle property 'mavenRepoPassword'")
```

Comment:
Credentials provider can be built in a way to expose a helpful message with hints about where the values should come from.